### PR TITLE
reclaim space of sda3 when auto-update disabled

### PR DESCRIPTION
**Pull request by hegao12**
When the auto-update of COS is disabled, one of
the two root partitions will not be used. In order
to reclaim the space of the unused root partition,
`extend-oem.go` is changed into `handle-disk-layout.go`
that modifies the partitions on the disk.

Since the stateful partition is adjacent to sda3 (rootA),
we will change the priority of the partitions to make the
system boot from sda5 (rootB), and then shrink sda3 to
reclaim the space it has.

After shrinking sda3, we need to reboot the system
so that it can read the new partition table. And this needs
to happend before the `resize-disk` step to prevent the
stateful partition being extended. Therefore, `startup.sh`
is now divided into two parts, `prepare()` and `build()`.
`prepare()` runs before the `resize-disk` step in the Daisy
workflow and `build()` runs after that. And the Daisy
workflow is changed accordingly.

If the space of sda3 is reclaimed, 2046MB will be freed to
use by the OEM partition and the stateful partition. So the
limit of disk-size-gb is changed as well.

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -227,6 +227,7 @@ dpkg_src(
     url = "http://snapshot.debian.org/archive",
 )
 
+
```
Comment (vaibhav-rustagi) [inline WORKSPACE]: nit: Can delete this change.

File: cmd/finish_image_build.go [hunk 2]
```diff
@@ -164,7 +164,10 @@ func (f *FinishImageBuild) loadConfigs(files *fs.Files) (*config.Image, *config.
 func validateOEM(buildConfig *config.Build) error {
 	// The default size of a COS image (imgSize) is assumed to be 10GB.
 	const imgSize uint64 = 10
-	var sizeErrorMsg string
+	// If auto-update is disabled, 2046MB will be reclaimed.
```
Comment (vaibhav-rustagi) [inline cmd/finish_image_build.go:167]: Can we also add why 2 MB is not reclaimed?

File: tools/partutil/handle_partition_table.go [hunk 3]
```diff
@@ -163,3 +167,73 @@ func ReadPartitionStart(disk string, partNumInt int) (uint64, error) {
...
+	var startSector uint64
+
+	// edit partition table.
+	table, err = ParsePartitionTable(table, partName, true, func(p *PartContent) {
+		startSector = p.Start
+		oldSize = p.Size
+		p.Size = minSize
+	})
+	if err != nil {
+		return 0, fmt.Errorf("error when editing partition table of %q, "+
```
Comment (vaibhav-rustagi) [inline tools/partutil/handle_partition_table.go:212]: The API name says `ParsePartitionTable`. Are you editing the partition table?

File: tools/handle_disk_layout.go [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+
+// HandleDiskLayout changes the partitions on a COS disk.
+// If the auto-update is disabled, it will shrink sda3 to reclaim the space.
+// If the OEM partition needs to be extended,
+// It also moves the OEM partition and the stateful partition by a distance relative
+// to a start point. If sda3 is shrinked, the start point is the end of shrinked sda3.
+// Otherwise, start point will be the original start of the stateful partition.
+// The stateful partition will be moved to leave enough space for the OEM partition,
+// and the OEM partition will be moved to the start point.
+// Finally OEM partition will be resized to 1 sector before the new stateful partition.
```
Comment (vaibhav-rustagi) [inline tools/handle_disk_layout.go:34]: What exactly is this case?

File: tools/handle_disk_layout.go [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+// OEMSize can be the number of sectors (without unit) or size like "3G", "100M", "10000K" or "99999B".
+// If no need to extend the OEM partition, oemSize=0.
+func HandleDiskLayout(disk string, statePartNum, oemPartNum int, oemSize string, reclaimSDA3 bool) error {
```
Comment (vaibhav-rustagi) [inline tools/handle_disk_layout.go:38]: It's a very long function. Is it possible to break it down into smaller pieces?
Also, the API is dealing with OEM partition. I was wondering why it is given a generic name?

**Review by rkolchmeyer — APPROVED** [review]

**Pull request merged**